File: pulsar-skeleton/src/main/java/ai/platon/pulsar/crawl/parse/html/FieldCollection.java
package ai.platon.pulsar.crawl.parse.html;

import java.util.HashMap;

/**
 * Created by vincent on 17-8-3.
 * Copyright @ 2013-2017 Platon AI. All rights reserved
 */
public class FieldCollection extends HashMap<CharSequence, CharSequence> {
    private String name;
    private int required = 0;
    private int loss = 0;

    public FieldCollection() {

    }

    public FieldCollection(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
        put("name", name);
    }

    public int getRequired() {
        return required;
    }

    public void setRequired(int required) {
        this.required = required;
    }

    public void increaseRequired(int count) {
        this.required += count;
    }

    public int getLoss() {
        return loss;
    }

    public void setLoss(int loss) {
        this.loss = loss;
    }

    public void loss(int loss) {
        this.loss += loss;
    }

    public FieldCollection assertContains(String key, String value) {
        assert (value.equals(get(key).toString()));
        // If (!equals) throw Exception
        return this;
    }

    public FieldCollection assertContainsKey(String... keys) {
        for (String key : keys) {
            assert (containsKey(key));
            // If (!equals) throw Exception
        }
        return this;
    }

    public FieldCollection assertContainsValue(String... values) {
        for (String value : values) {
            assert (containsValue(value));
            // If (!equals) throw Exception
        }
        return this;
    }

    public FieldCollection assertContains(String message, String key, String value) {
        if (!value.equals(get(key).toString())) {
            System.out.println(key + ", " + "value");
            System.out.println(message);
        }
        return this;
    }
}

File: pulsar-skeleton/src/main/kotlin/ai/platon/pulsar/crawl/parse/html/FieldCollection.kt
package ai.platon.pulsar.crawl.parse.html

import java.util.*

/**
 * Created by vincent on 17-8-3.
 * Copyright @ 2013-2017 Platon AI. All rights reserved
 *
 * TODO: bad to inherit from HashMap
 */
class FieldCollection : HashMap<String, String>() {
    var name: String? = null
        set(value) {
            field = value
            if (value != null) {
                put("name", value)
            }
        }

    var required = 0
    var loss = 0

    fun increaseRequired(count: Int) {
        required += count
    }

    fun loss(loss: Int) {
        this.loss += loss
    }

    fun assertContains(key: String, value: String): FieldCollection {
        assert(value == get(key).toString())
        // If (!equals) throw Exception
        return this
    }

    fun assertContainsKey(vararg keys: String): FieldCollection {
        for (key in keys) {
            assert(containsKey(key))
            // If (!equals) throw Exception
        }
        return this
    }

    fun assertContainsValue(vararg values: String): FieldCollection {
        for (value in values) {
            assert(containsValue(value))
            // If (!equals) throw Exception
        }
        return this
    }

    fun assertContains(message: String, key: String, value: String): FieldCollection {
        return this
    }
}